### PR TITLE
Add support for identities from W3C Spec in server. Closes #73

### LINKED ISSUE #73: Add inferred_from field to Annotations in the server
It might be better to record them in the opposite way so you don't have to keep the existing annotations in sync when one is deleted. In other words, maybe adding a field like `inferred_from` or something. This field could be added by the recognition API, btw.

_Originally posted by @brollb in https://github.com/symbench/scm/pull/63#discussion_r650077506_

### PR DESCRIPTION
**Pull request by umesh-timalsina**

### CODE REVIEW

**Comment by umesh-timalsina** [comment]
It looks like you can add inferred fields using [`identities`](https://www.w3.org/TR/annotation-model/#other-identities) with the W3C spec. c69fa65 Adds such relationship.

File: circuit_recognizer/circuit_recognizer/annotations.py [hunk 1]
```diff
@@ -24,6 +24,8 @@ class Annotation(BaseModel):
         "type": "Agent",
         "name": "circuit_recognizer.download_and_annotate"
     }
+    cannonical: Optional[str] = None
```
Comment (brollb) [inline circuit_recognizer/circuit_recognizer/annotations.py:27]: I believe this should be `canonical` (two total "n"s), right?

Comment (umesh-timalsina) [reply circuit_recognizer/circuit_recognizer/annotations.py]: Oops My Bad, can even make an excuse for British spelling here. :)

File: circuit_recognizer/circuit_recognizer/annotations.py [hunk 2]
```diff
@@ -102,11 +121,17 @@ def add_port(self, annotation):
...
-        for port in copy.ports:
-            port.id = str(uuid4())
+
+        for port_copy, original_port in zip(copy.ports, self.ports):
+            port_copy.id = str(uuid4())
+            if infer:
+                port_copy.cannonical = original_port.id
+
+        if infer:
+            copy.cannonical = self.id
```
Comment (brollb) [inline circuit_recognizer/circuit_recognizer/annotations.py:134]: typo

File: circuit_recognizer/circuit_recognizer/annotations.py [hunk 3]
```diff
@@ -102,11 +121,17 @@ def add_port(self, annotation):
...
+    def copy(self, infer=False):
         copy = deepcopy(self)
         copy.id = str(uuid4())
-        for port in copy.ports:
-            port.id = str(uuid4())
+
+        for port_copy, original_port in zip(copy.ports, self.ports):
+            port_copy.id = str(uuid4())
+            if infer:
+                port_copy.cannonical = original_port.id
```
Comment (brollb) [inline circuit_recognizer/circuit_recognizer/annotations.py:131]: typo in canonical

**Review by brollb — CHANGES_REQUESTED** [review]
Overall, this looks great. One minor stylistic suggestion: instead of adding a flag to `copy`, it might be nicer to always copy the canonical representation. It isn't obvious what `infer` does at a glance and it seems that recording the canonical representation is always the intended behavior.

File: circuit_recognizer/circuit_recognizer/annotations.py [hunk 4]
```diff
@@ -61,12 +63,29 @@ def rotate(self, rotation, center=None):
...
         _, _, width, height = self.bounding_box()
+        if self.has_canonical_representation():
+            self.add_relationship(f'translation(xy)=pixel:{x},{y}')
         self.target['selector']['value'] = f'xywh=pixel:{x},{y},{width},{height}'
 
+    def has_canonical_representation(self):
+        return self.canonical is not None
+
+    def add_relationship(self, relation):
+        if not self.canonical:
```
Comment (brollb) [inline circuit_recognizer/circuit_recognizer/annotations.py:80]: Just using `assert` would probably be a little more concise but not a huge deal

Comment (umesh-timalsina) [reply circuit_recognizer/circuit_recognizer/annotations.py:80]: Yeah, in this case `ValueError` makes more sense IMO.

Comment (brollb) [reply circuit_recognizer/circuit_recognizer/annotations.py:80]: Hmm... I am not sure I follow. Wouldn't a value error imply there is something wrong with the value of `relation`? In this case, the problem seems to be that the method is invalid for instances without the appropriate state. (For reference, [here](https://docs.python.org/3/library/exceptions.html#ValueError) is the official description for ValueError I am looking at.)

**Comment by umesh-timalsina** [comment]
> Overall, this looks great. One minor stylistic suggestion: instead of adding a flag to `copy`, it might be nicer to always copy the canonical representation. It isn't obvious what `infer` does at a glance and it seems that recording the canonical representation is always the intended behavior.

Two comments on this:
1. Isn't it the case that copy is generally used to create the exact replica of the object under consideration? 
2. In this case we might have to consider that `copy` is also a method in pydantic's `BaseModel`, which is overriden here. This might be a case where we don't want the use super's copy but having a flag which by default is False (for inference) gives us almost the same behavior.

**Comment by brollb** [comment]
Looks great!

**Review by brollb — APPROVED** [review]

**Pull request merged**